File: 백준/Gold/2573. 빙산/빙산.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int N, M;

int dx[4] = { 0, 0, -1, 1 };
int dy[4] = { -1, 1, 0, 0 };

void dfs(int x, int y, vector<vector<int>>& map, vector<vector<bool>>& visited) {
	visited[x][y] = true;

	for (int dir = 0; dir < 4; dir++) {
		int nx = x + dx[dir];
		int ny = y + dy[dir];

		if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
			if (!visited[nx][ny] && map[nx][ny] > 0)
				dfs(nx, ny, map, visited);
		}
	}
}

int countIcebergs(vector<vector<int>>& map) {
	vector<vector<bool>> visited(N, vector<bool>(M, false));
	int count = 0;

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if (!visited[i][j] && map[i][j] > 0) {
				dfs(i, j, map, visited);
				count++;
			}
		}
	}

	return count;
}

void melt(vector<vector<int>>& map) {
	vector<vector<int>> temp = map;

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if (map[i][j] > 0) {
				int sea = 0;

				for (int dir = 0; dir < 4; dir++) {
					int nx = i + dx[dir];
					int ny = j + dy[dir];
					
					if (nx >= 0 && nx < N && ny >= 0 && ny < M && map[nx][ny] == 0)
						sea++;
				}

				temp[i][j] = max(0, map[i][j] - sea);
			}
		}
	}

	map = temp;
}

int main() {
	// Step 1: 입력
	cin >> N >> M;

	vector<vector<int>> map(N, vector<int>(M));
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++)
			cin >> map[i][j];
	}

	// Step 2
	int year = 0;

	while (true) {
		int count = countIcebergs(map);
		
		if (count == 0) {
			cout << 0 << endl;
			break;
		}

		if (count >= 2) {
			cout << year << endl;
			break;
		}

		melt(map);
		year++;
	}

	return 0;
}

File: 백준/Gold/2573. 빙산/빙산.py
import sys
sys.setrecursionlimit(10**7)

# 방향 벡터 상, 하, 좌, 우
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

# DFS로 연결 요소(빙산 덩어리) 탐색
def dfs(x, y, iceberg_map, visited, n, m):
    visited[x][y] = True
    for dir in range(4):
        nx, ny = x + dx[dir], y + dy[dir]
        if 0 <= nx < n and 0 <= ny < m:
            if iceberg_map[nx][ny] > 0 and not visited[nx][ny]:
                dfs(nx, ny, iceberg_map, visited, n, m)

# 빙산 덩어리 개수 세기
def countIcebergs(iceberg_map, n, m):
    visited = [[False] * m for _ in range(n)]
    count = 0
    for i in range(n):
        for j in range(m):
            if iceberg_map[i][j] > 0 and not visited[i][j]:
                dfs(i, j, iceberg_map, visited, n, m)
                count += 1
    return count

# 빙산 녹이기
def melt(iceberg_map, n, m):
    temp = [row[:] for row in iceberg_map]
    for i in range(n):
        for j in range(m):
            if iceberg_map[i][j] > 0:
                sea = 0
                for dir in range(4):
                    nx, ny = i + dx[dir], j + dy[dir]
                    if 0 <= nx < n and 0 <= ny < m and iceberg_map[nx][ny] == 0:
                        sea += 1
                temp[i][j] = max(0, iceberg_map[i][j] - sea)
    return temp

# 메인 함수
def main():
    n, m = map(int, input().split())
    iceberg_map = [list(map(int, input().split())) for _ in range(n)]
    year = 0

    while True:
        cnt = countIcebergs(iceberg_map, n, m)
        if cnt == 0:
            print(0)
            return
        if cnt >= 2:
            print(year)
            return
        iceberg_map = melt(iceberg_map, n, m)
        year += 1

if __name__ == "__main__":
    main()
